File: app/src/main/java/be/digitalia/fosdem/viewmodels/PersonsViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.model.Person;

public class PersonsViewModel extends AndroidViewModel {

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final LiveData<PagedList<Person>> persons
			= new LivePagedListBuilder<>(appDatabase.getScheduleDao().getPersons(), 100)
			.build();

	public PersonsViewModel(@NonNull Application application) {
		super(application);
	}

	public LiveData<PagedList<Person>> getPersons() {
		return persons;
	}
}

File: app/src/main/kotlin/be/digitalia/fosdem/viewmodels/PersonsViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.cachedIn
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.model.Person
import be.digitalia.fosdem.paging.toAutoCloseable
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

@HiltViewModel
class PersonsViewModel @Inject constructor(scheduleDao: ScheduleDao) : ViewModel() {

    val persons: Flow<PagingData<Person>> = Pager(PagingConfig(20)) {
        scheduleDao.getPersons()
            .also { addCloseable("persons", it.toAutoCloseable()) }
    }.flow.cachedIn(viewModelScope)
}